File: DEV3600/DEV360/LESSON_3/auctionsapp/src/main/scala/solutions/AuctionsApp.scala
package solutions
import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.apache.spark.SparkConf
import java.lang.Math

object AuctionsApp {
  def main(args: Array[String]) {
    val conf = new SparkConf().setAppName("AuctionsApp")
    val sc = new SparkContext(conf)
    val usrhome = System.getenv("HOME")
    val aucFile = usrhome.concat("/data/auctiondata.csv")

    // Map input variables                
    val auctionid = 0
    val bid = 1
    val bidtime = 2
    val bidder = 3
    val bidderrate = 4
    val openbid = 5
    val price = 6
    val itemtype = 7
    val daystolive = 8

    // Load the data
    val auctionRDD = sc.textFile(aucFile).map(line => line.split(",")).cache()

    // Total number of bids across all auctions
    val totalbids = auctionRDD.count()

    // Total number of items
    val totalitems = auctionRDD.map(line => line(auctionid)).distinct().count()

    // Max, min and avg number of bids
    val bids_auctionRDD = auctionRDD.map(x => (x(auctionid), 1)).reduceByKey((x, y) => x + y)
    val maxbids = bids_auctionRDD.map(x => x._2).reduce((x, y) => Math.max(x, y))
    val minbids = bids_auctionRDD.map(x => x._2).reduce((x, y) => Math.min(x, y))
    val avgbids = totalbids / totalitems

    // Print output to the console
    println("total bids across all auctions: %s ".format(totalbids))
    println("total number of distinct items: %s ".format(totalitems))
    println("Max bids across all auctions: %s ".format(maxbids))
    println("Min bids across all auctions: %s ".format(minbids))
    println("Avg bids across all auctions: %s ".format(avgbids))
  }
}

File: DEV3600/DEV360/LESSON_3/auctions_app.py
from pyspark import SparkContext,SparkConf

# Define variables
conf = SparkConf().setAppName("auctions_app")
sc = SparkContext(conf=conf)

# Map input variables                
auctionid = 0
bid = 1
bidtime = 2
bidder = 3
bidderrate = 4
openbid = 5
price = 6
itemtype = 7
daystolive = 8

# Load the data
auctionRDD = sc.textFile("/user/user01/data/auctiondata.csv").map(lambda line: line.split(",")).cache()

# Total number of bids across all auctions
totalbids = auctionRDD.count()

# Total number of items
totalitems = auctionRDD.map(lambda x: x[auctionid]).distinct().count()

# Max, min and avg number of bids
bids_auctionRDD = auctionRDD.map(lambda x: (x[auctionid], 1)).reduceByKey(lambda x, y: x + y)
maxbids = bids_auctionRDD.map(lambda x: x[bid]).reduce(max)
minbids = bids_auctionRDD.map(lambda x: x[bid]).reduce(min)
avgbids = totalbids / totalitems

# Print output to the console
print "Total bids across all auctions: %d " %(totalbids)
print "Total number of distinct items: %d " %(totalitems)
print "Max bids across all auctions: %d " %(maxbids)
print "Min bids across all auctions: %d " %(minbids)
print "Avg bids across all auctions: %d " %(avgbids)
print "DONE"                                                         
